keylime/keylime_agent.py

```diff
@@ -723,7 +733,24 @@ def main():
     if keylime_ca == "default":
         keylime_ca = os.path.join(config.WORK_DIR, 'cv_ca', 'cacert.crt')
 
-    server = CloudAgentHTTPServer(serveraddr, Handler, agent_uuid, contact_ip)
+    # Now that root-owned files are opened drop privileged
+    # If 'run_as' is not available keep running with high privileges
+    if os.getuid() == 0:
+        run_as = config.get('cloud_agent', 'run_as', fallback='')
+        if run_as != '':
+            for p, recurse in [("", False),
+                               ("tpmdata.yml", False),
+                               ("secure", True),
+                               ("cv_ca", True)]:
```

Comment:
This might makes sense since `cv_ca` is the default location when searching for certiticates but it is also possible that this directory won't exist. Is this really intentional? I am asking because the other two locations are created by the agent itself but that does not apply to `cv_ca`.

Reply:
I wasn't sure what creates this directory. How should it be handled?

Reply:
It is created by verifier which uses it to store generated certificates, as described here.
https://github.com/keylime/keylime/blob/3590e21b7e4b48a2023aadf2486b116ef5be2375/keylime.conf#L169
While agent is using cacert and client-cert from this directory (not in my case as I am copying them to other location) I think it should not own that whole directory. Personally, I would not list it as these certificates should exist before agent starts and therefore it is up to admin to configure permissions and agent accordingly.
@THS-on thoughts?

Reply:
Is there a possibility of not doing this?

IMHO this is a security hole. If some user manage to put something there, it will be now automatically owned by the agent.  Is very good that the current code is picky and is selecting by name the elements that want to change, but still we can crate link and start making the agent to chown stuff outside the WORK_DIR.

Also, as commented here, this makes the verifier and the agent not co-installable. For me this is OK, but is a signal of that there is something wrong in the design.

Ideally the agent should not take any action, here nor outside, as the less code the more secure will be (and dumb). It should check if some action can be done, and if no die!

We need to document all the preparations required to work the agent in different scenario. For me is totally OK the idea suggesting that if the agent cannot mount the tmpfs, is game over.  If it is not root, the admin needs to prepare the system so the agent can mount the tmpfs (maybe adjusting fstab, or better with a systemd mount service).  The same for the certificates: if the agent cannot read them, something did go wrong during the installation, so die.

Does this makes sense? Am I missing something important that is against this idea?

Reply:
I agree with @aplanas we should not change permissions of arbitrary files. My proposed changes are:
1. move the privilege drop before  `# Instanitate TPM class` (line 640) (Note that dmidecode should run before because I think it requires root)
2. Fail if tpmfs is not there. Do not try to create it if the agent does not run as root.
3. Fail if the Keylime CA certificate is not readable

Reply:
Why do we have tmpfs there? Is this to be able to unmount and forget the files under this directory, so an equivalent of an `rm -rf <dir>` at shutdown that wouldn't need root rights and therefore be much more flexible?

> but still we can crate link and start making the agent to chown stuff outside the WORK_DIR.

I tried to prevent exactly this by using `follow_symlinks = False`.
```
os.chown(rootpath, uid, gid, follow_symlinks = False)
```

os.walk doesn't follow links by default.
```
os.walk(top[, topdown=True[, onerror=None[, followlinks=False]]])
```

Reply:
> Why do we have tmpfs there?

Do not take my word, but I think that is optional. The admin can create a `secure` directory beforehand, and will be used by the agent. Only if is not present will create one as a `tmpfs`. The advantage is that is not persistent, so will be gone always, as you commented.

> I tried to prevent exactly this by using `follow_symlinks = False`.

Yes, you are being super careful here. I liked a lot that you are enumerating the elements and not doing a wild `chown`.

But the basic idea remains, I think. If somehow a file is dropped here under with my "aplanas:user", it will be `chown`ned by the agent. This is the case for `cv_ca`, that can be there because was created by the verifier!

My suggestion is more like a principle: the agent should do the least amount of work possible, and IMHO the `chown` is something that we can skip here with preparation and documentation. I am not a security knowledge developer, but my intuition tells me that a `chown` is something that a hacker will find interesting.